### PR TITLE
fix(pubkeyimporter): jS error when using Enter as text with incorrect access type

### PR DESCRIPTION
**Pull request by bruce1971**
Handling the error when an uncompressed public key is given to address types P2WSH or P2SH-P2WSH

ISSUES: https://github.com/unchained-capital/caravan/issues/164

Demo video:
https://user-images.githubusercontent.com/76526363/103631741-a97ee180-4f43-11eb-93dd-10a23ecdcb52.mp4

Thank for taking a look & sending feedbacks my way :)

### CODE REVIEW

**Comment by bucko13** [comment]
Thanks for the contribution as well as for the very helpful walkthrough video! This seems to solve the problem correctly however my feeling is that this might be more useful as a validation utility in [unchained-bitcoin](https://github.com/unchained-capital/unchained-bitcoin). Two thoughts that come to mind is that this could either be part of the existing `validatePublicKey`, which you used in your implementation, and have `addressType` as an optional param, or alternatively what might be a little more "sanitary" is call this something like `validatePublicKeyForAddressType` which does as you've done. 

Also note that unchained-bitcoin does have a tool to check if the key is compressed [isKeyCompressed](https://github.com/unchained-capital/unchained-bitcoin/blob/master/src/keys.js#L455) which you should be able to reuse.

**Comment by bruce1971** [comment]
Thanks @bucko13 for taking a look at my PR, appreciate!
Yes, its a fair point you make! Let me have a look at the `validatePublicKey` function in the unchained-bitcoin repo, and see if I can come up with a cleaner implementation! Also happy to hear that `isKeyCompressed` exists!

**Comment by bruce1971** [comment]
Hi guys!

As you suggested @bucko13, I went ahead and slightly extended the `validatePublicKey()` function (see this PR) to accept addressType as a parameter.

Then I further modified this PR to make use of the updated functionality of `validatePublicKey()`.

Thank you for taking a look! Let me know if you have any questions/feedback :)

File: src/components/CreateAddress/PublicKeyImporter.jsx
Comment (bucko13) [inline src/components/CreateAddress/PublicKeyImporter.jsx:349]: I don't think we need to do this twice (running it in the message too). Maybe run the validation before return and if it has length then pass that to `Conflict`

File: src/components/CreateAddress/PublicKeyImporter.jsx
Comment (bucko13) [inline src/components/CreateAddress/PublicKeyImporter.jsx:61]: Can we avoid the nested if here? Maybe just do an `&& !validatePublicKey(...)` since it doesn't look like the actual error value is being used.

File: src/components/CreateAddress/AddressGenerator.jsx
Comment (bucko13) [inline src/components/CreateAddress/AddressGenerator.jsx:9]: Just a note that this PR will also require a version bump to `unchained-bitcoin` before it can be merged in.

File: src/components/CreateAddress/AddressGenerator.jsx
Comment (bucko13) [inline src/components/CreateAddress/AddressGenerator.jsx:48]: this is just a personal preference/styling thing, but it might look cleaner to destructure in the filter callback 

```suggestion
    return Object.values(publicKeyImporters).filter(({ publicKey, finalized }) => 
      finalized && !validatePublicKey(publicKey, addressType)
    ).length;
```
(I haven't tested this so please confirm it works)

**Review by bucko13 — CHANGES_REQUESTED** [review]
Looking good! A couple of comments, and obviously will need to wait for the unchained-bitcoin updates to be published, but thanks for taking the time to address the comments!

**Comment by waldenraines** [comment]
unchained-bitcoin version `0.1.3` has been released to be used in this PR (i.e. includes https://github.com/unchained-capital/unchained-bitcoin/pull/25).

File: src/components/CreateAddress/PublicKeyImporter.jsx
Comment (dhruvbansal) [inline src/components/CreateAddress/PublicKeyImporter.jsx:186]: This modification is necessary because we must pass `addressType` to the `TextPublicKeyImporter`.

Our suspicion is that none of the *other* changes in this file are necessary.

**Review by dhruvbansal — CHANGES_REQUESTED** [review]
We expected the changes to `validatePublicKey` to only effect the `TextPublicKeyImporter`, since that's the only public key import method through which we could trigger a mismatch between compressed/uncompressed public keys and address types.

As a result, we believe the following two minimal changes (both of which you've already made) are required to fix this bug:

* The changes to `TextPublicKeyImporter`
* The change on line 186 of `PublicKeyImporter`

Other changes in `PublicKeyImporter` and `AddressGenerator` may be superfluous.  **Can you please test this fix with just the two minimal changes above?** If everything works as expected, then we'd prefer to merge just these minimal changes.

One might argue that, since the signature of the `validatePublicKey` method (from `unchained-bitcoin`) has changed to include `addressType` as a second parameter, that all usages of `validatePublicKey` should be updated to include the additional `addressType`.

This is true from a consistency standpoint though we believe *not* including this additional parameter won't actually cause any bugs since it's only the `TextPublicKeyImporter` that allows creating the mismatch in the first place.  We have some plans to rewrite this part of the code, so we're OK being a little inconsistent right now in favor of making the minimal changes above.

Thanks!

**Comment by bruce1971** [comment]
Hi guys,

When tackling this bug, I realized that there are actually 2 separate cases to fix (explained in the video!).

CASE 1 - First select address Type P2WSH, then import uncompressed Public Key.
CASE 2 - Whilst on the default P2SH, import uncompressed Public Key, THEN change address selector to P2WSH.

Yes CASE 1 can be fixed with the two minimal changes @dhruvbansal and the ValidatePublicKey change in the unchained-bitcoin package.

However to fix CASE 2, more code is needed… and that’s essentially what the rest of the code is for. Yes I agree, that this additional code can be hacky/unclean. The complication for Case 2 arrises, when the uncompressed PK is imported under P2SH.. At this point is is valid, and `finalized`=true is set. Then however, when the addressType is changed to P2WSH, the setup becomes invalid… but `finalized` still remains true. One reason for this complication, is that AddressTypePicker is just a simple component that only changes the addressType state, but not the already imported Public Keys…

Anyways, if you guys say you prefer to fix Case 1 first and leave Case 2 open for now, as this part of the code might be refactored, I’m fine with that too! Let me know, and I’ll finalize the changes and we can merge!

**Comment by bruce1971** [comment]
any thoughts on my comment above folks? @bucko13 @dhruvbansal 
If you want to keep it simple, and only fix CASE 1, let me know and ill make the adjustments
Thanks

**Comment by bucko13** [comment]
Hey @bruce1971- I think that makes sense to me. It's likely that we'll have to do some broader cleanup and refactoring to make the type of problem you're solving slightly less convoluted. In any case, I'll let @dhruvbansal have the final word. In the meantime, the other styling fixes should still be applicable and the unchained-bitcoin dependency will still need to be bumped.

**Review by dhruvbansal — DISMISSED** [review]

**Comment by bucko13** [comment]
Hey @bruce1971 just noting that @dhruvbansal is good to go with the approach you've taken. I think we'll leave some of the larger structural issues up for another more targeted task. Better to fix these bugs you've identified (and fixed) now rather than later. Just need to address the code comments and bump unchained-bitcoin and this should be ready to go in! Sorry for the delays in getting to this.

File: src/components/CreateAddress/PublicKeyImporter.jsx
Comment (bruce1971) [reply src/components/CreateAddress/PublicKeyImporter.jsx:349]: fair, dont need to run this twice

but there is the case when the error is `Public key cannot be blank.` which has length... so need to use a slightly different criteria there..

**Comment by bruce1971** [comment]
Hey @bucko13, added all your requested changes. Tested them manually and it seems to work as expected in the app.
Let me know if we're good to merge, or you want any other changes. Cheers!

File: src/components/CreateAddress/PublicKeyImporter.jsx
Comment (bucko13) [reply src/components/CreateAddress/PublicKeyImporter.jsx]: Awesome. Good call and works for me.

**Review by bucko13 — APPROVED** [review]
Many thanks @bruce1971. Tested and looks good to me. Code looks like it checks out too!

**Pull request merged**
